### PR TITLE
ARROW-7670: [Python][Dataset] More ergonomical API

### PR DESCRIPTION
**Pull request by kszucs**
- introduced `ds.field()`, `ds.scalar()`
- overloaded comparison and binary operators 
- added `Expression.isin()` and `Expression.is_valid()`
- removed ScannerBuilder in favor of direct `Scanner` instantiation using keyword arguments
- added convenience methods for Dataset: `to_table()`, `to_batches()` and `scan()`
- added support for string scalar
- removed `InMemoryScanTask` and `DefaultPartitioning`

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
Is there a reason you renamed Expression to Expr? Personally, I don't really like such abbreviations in class names in general.

**Comment by jorisvandenbossche** [comment]
But nice interface!

**Comment by kszucs** [comment]
I found it too verbose, besides that I'm pretty used to it because of ibis. Besides that we have another expression [interface](https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/expression.h#L37) which IMO is going to be the base in the future - it also uses abbreviations.

File: python/pyarrow/_dataset.pyx [hunk 1]
```diff
@@ -585,17 +588,17 @@ cdef class Source:
         return pyarrow_wrap_schema(self.source.schema())
 
     @property
-    def partition_expression(self):
+    def partition_Expr(self):
```
Comment (kszucs) [inline python/pyarrow/_dataset.pyx]: Renaming error.

**Comment by kszucs** [comment]
TODO: add the two new functions to the apidoc

**Review by bkietz — CHANGES_REQUESTED** [review]
Please revert the rename `Expression -> Expr` and write a separate JIRA for it. I'm not opposed to renaming the class, but I do think that Python, C++ and R should all match on the name. If we're going to rename like this then we should rename in all the languages in sync

**Comment by kszucs** [comment]
I thought that R already uses Expr.

File: python/pyarrow/dataset.py [hunk 2]
```diff
@@ -328,3 +326,37 @@ def dataset(sources, filesystem=None, partitioning=None, format=None):
...
+
+    Returns
+    -------
+    field_expr : FieldExpression
+    """
+    return FieldExpression(name)
+
+
+def scalar(value):
+    """Expressionession representing a scalar value.
```
Comment (bkietz) [inline python/pyarrow/dataset.py]:
```suggestion
    """Expression representing a scalar value.
```

File: python/pyarrow/_dataset.pyx [hunk 3]
```diff
@@ -816,163 +833,75 @@ cdef class ScanTask:
...
     """
 
     cdef:
-        shared_ptr[CScannerBuilder] wrapped
-        CScannerBuilder* builder
+        shared_ptr[CScanner] wrapped
+        CScanner* scanner
 
-    def __init__(self, Dataset dataset not None, MemoryPool memory_pool=None):
+    def __init__(self, Dataset dataset, list columns=None, Expression filter=None,
```
Comment (bkietz) [inline python/pyarrow/_dataset.pyx]: Style failure https://github.com/apache/arrow/pull/6268/checks?check_run_id=407444654#step:5:55
```
autopep8 --global-config=.flake8.cython -i pyarrow/_dataset.pyx
```

File: python/pyarrow/_dataset.pyx [hunk 4]
```diff
@@ -639,28 +624,28 @@ cdef class FileSystemSource(Source):
         ----------
         schema : Schema
             Schema for resulting Source
-        source_partition : Expression
+        source_partition : Expr
```
Comment (bkietz) [inline python/pyarrow/_dataset.pyx]:
```suggestion
        source_partition : Expression
```

File: python/pyarrow/_dataset.pyx [hunk 5]
```diff
@@ -639,28 +624,28 @@ cdef class FileSystemSource(Source):
...
         file_format : FileFormat
         filesystem : FileSystem
             FileSystem which will be explored to discover data files.
         paths_or_selector : Union[FileSelector, List[FileStats]]
             The file stats object can be queried by the
             filesystem.get_target_stats method.
-        partitions : List[Expression]
+        partitions : List[Expr]
```
Comment (bkietz) [inline python/pyarrow/_dataset.pyx]:
```suggestion
        partitions : List[Expression]
```

File: python/pyarrow/_dataset.pyx [hunk 6]
```diff
@@ -733,26 +718,58 @@ cdef class Dataset:
...
-            For memory allocations, if required. If not specified, uses the
-            default pool.
+        columns : list of str, default None
+            List of columns to project. Order and duplicates will be preserved.
+            The goal is to avoid loading, copying, and deserializing columns
+            that will not be required further down the compute chain.
+            By default all of the available columns are projected, raises
+            an exception if any of the referenced column names does not exist
+            in the dataset's Schema.
+        filter : Expr, default None
```
Comment (bkietz) [inline python/pyarrow/_dataset.pyx]:
```suggestion
        filter : Expression, default None
```

File: python/pyarrow/_dataset.pyx [hunk 7]
```diff
@@ -816,163 +833,75 @@ cdef class ScanTask:
...
+    dataset : Dataset instance
         The dataset to scan.
+    columns : list of strings, default None
+        List of projectable columns during the scan operation. Order and
+        duplicates will be preserved.
+        This subset will be passed down to Sources and corresponding data
+        fragments. The goal is to avoid loading, copying, and deserializing
+        columns that will not be required further down the compute chain.
+        By default all available columns will be projected.
+    filter : Expr, default None
```
Comment (bkietz) [inline python/pyarrow/_dataset.pyx]:
```suggestion
    filter : Expression, default None
```

**Review by bkietz — CHANGES_REQUESTED** [review]
a few typos

File: python/pyarrow/_dataset.pyx [hunk 8]
```diff
@@ -806,173 +884,88 @@ cdef class ScanTask:
...
 
-    It is used to pass information, notably a potential filter expression and a
-    subset of columns to materialize.
+    A scanner is the class that glues the scan tasks, data fragments and data
+    sources together.
 
     Parameters
     ----------
-    dataset : Dataset
-        The dataset to scan.
```
Comment (jorisvandenbossche) [inline python/pyarrow/_dataset.pyx]: it still takes the dataset as first argument?

File: python/pyarrow/_dataset.pyx [hunk 9]
```diff
@@ -733,30 +721,120 @@ cdef class Dataset:
...
+            default pool.
+
+        Returns
+        -------
+        scan_tasks : iterator of ScanTask
         """
-        Begin to build a new Scan operation against this Dataset.
+        scanner = Scanner(self, columns=columns, filter=filter,
+                          memory_pool=memory_pool)
+        return scanner.scan()
```
Comment (jorisvandenbossche) [inline python/pyarrow/_dataset.pyx]: An option is also to return the `scanner` object itself?

File: python/pyarrow/_dataset.pyx [hunk 10]
```diff
@@ -733,30 +721,120 @@ cdef class Dataset:
...
+                          memory_pool=memory_pool)
+        for task in scanner.scan():
+            for batch in task.execute():
+                yield batch
+
+    def to_table(self, columns=None, filter=None, use_threads=True,
+                 MemoryPool memory_pool=None):
+        """Read the dataset to an arrow table.
+
+        Note, that this method reads all the selected data from the dataset
```
Comment (jorisvandenbossche) [inline python/pyarrow/_dataset.pyx]:
```suggestion
        Note that this method reads all the selected data from the dataset
```

File: python/pyarrow/dataset.py [hunk 11]
```diff
@@ -328,3 +326,37 @@ def dataset(sources, filesystem=None, partitioning=None, format=None):
     discovery = sources[0]
     inspected_schema = discovery.inspect()
     return Dataset([discovery.finish()], inspected_schema)
+
+
+def field(name):
+    """References a named column of the dataset.
+
+    Stores only the field's name, type and other information is known only when
```
Comment (jorisvandenbossche) [inline python/pyarrow/dataset.py]:
```suggestion
    Stores only the field's name. Type and other information is known only when
```

File: python/pyarrow/includes/libarrow_dataset.pxd [hunk 12]
```diff
@@ -300,10 +300,6 @@ cdef extern from "arrow/dataset/api.h" namespace "arrow::dataset" nogil:
     cdef cppclass CPartitioningFactory "arrow::dataset::PartitioningFactory":
         pass
 
-    cdef cppclass CDefaultPartitioning \
-            "arrow::dataset::DefaultPartitioning"(CPartitioning):
-        CDefaultPartitioning()
-
```
Comment (jorisvandenbossche) [inline python/pyarrow/includes/libarrow_dataset.pxd]: You can probably remove `InMemoryScanTask` from this file as well

File: python/pyarrow/_dataset.pyx [hunk 13]
```diff
@@ -1071,21 +1087,57 @@ cdef class Expression:
...
+        return ComparisonExpression(operator_mapping[op], self, other)
+
+    def __and__(self, other):
+        return _binop(AndExpression, self, other)
+
+    def __or__(self, other):
+        return _binop(OrExpression, self, other)
+
+    def is_valid(self):
+        return IsValidExpression(self)
```
Comment (jorisvandenbossche) [inline python/pyarrow/_dataset.pyx]: Is this about being not null?

File: python/pyarrow/_dataset.pyx [hunk 14]
```diff
@@ -1071,21 +1087,57 @@ cdef class Expression:
...
+
+    def cast(self, type, bint safe=True):
+        return CastExpression(self, to=ensure_type(type), safe=safe)
+
+    def isin(self, values):
+        return InExpression(self, pa.array(values))
```
Comment (jorisvandenbossche) [inline python/pyarrow/_dataset.pyx]: Should this (and `cast` above) only be defined on a FieldExpression?

**Review by jorisvandenbossche — COMMENTED** [review]
Cool! Looking good ;)

I tested it on my previous taxi experiments, and this gives a nice improvement in usability (and for some reason I now also get the parallellisation)

**Comment by kszucs** [comment]
@jorisvandenbossche did you use python threading? Because the gil is released now in ScanTask.execute()

File: python/pyarrow/_dataset.pyx (hunk 14 above)
Comment (kszucs) [reply python/pyarrow/_dataset.pyx]: It makes sense on scalars, like `ds.scalar(1.0).cast('int64').isin([1, 2, 3])`
But indeed, in the future we'll need to restrict certain operations on specific expression types, like in ibis [1](https://github.com/ibis-project/ibis/blob/master/ibis/expr/api.py#L1085), [2](https://github.com/ibis-project/ibis/blob/master/ibis/expr/api.py#L1436), [3](https://github.com/ibis-project/ibis/blob/master/ibis/expr/api.py#L2590), etc.

File: python/pyarrow/includes/libarrow_dataset.pxd (hunk 12 above)
Comment (kszucs) [reply python/pyarrow/includes/libarrow_dataset.pxd]: Indeed

File: python/pyarrow/_dataset.pyx (hunk 9 above)
Comment (kszucs) [reply python/pyarrow/_dataset.pyx]: The scanner object can be easily instantiated by `Scanner(dataset, **kwargs)`

**Comment by jorisvandenbossche** [comment]
> @jorisvandenbossche did you use python threading? Because the gil is released now in ScanTask.execute()

I used Scanner.to_table, which was already releasing the gil if I looked correctly?

**Comment by kszucs** [comment]
Yeah to_table releases the GIL from the beginning, I have not changed that.

File: python/pyarrow/_dataset.pyx (hunk 13 above)
Comment (kszucs) [reply python/pyarrow/_dataset.pyx]: Added a short doctstring for that.

File: python/pyarrow/includes/libarrow_dataset.pxd (hunk 12 above)
Comment (kszucs) [reply python/pyarrow/includes/libarrow_dataset.pxd]: Removed

File: python/pyarrow/_dataset.pyx (hunk 8 above)
Comment (kszucs) [reply python/pyarrow/_dataset.pyx]: Yep, nice catch! Added the dataset parameter back to the docstring.

**Comment by kszucs** [comment]
@jorisvandenbossche updated

**Review by kszucs — COMMENTED** [review]
+1

**Pull request closed without merging**